### PR TITLE
Prune explicit nulls from client-side apply create

### LINKED ISSUE #123304: kubectl not idempotent when setting `null` values
### What happened?

As per [this article](https://home.robusta.dev/blog/stop-using-cpu-limits), I'm trying to remove CPU limits.

But the behaviour is different depending whether the ResourceQuota exists or not:

```
$ kubectl describe quota
No resources found in deleteme namespace.

$ kubectl apply -f - <<EOF
kind: ResourceQuota
apiVersion: v1
metadata:
  name: compute-resources
  namespace: deleteme
spec:
  hard:
    limits.cpu: null
EOF
resourcequota/compute-resources created

$ kubectl describe quota
Name:       compute-resources
Namespace:  deleteme
Resource    Used  Hard
--------    ----  ----
limits.cpu  0     0

$ kubectl apply -f - <<EOF
kind: ResourceQuota
apiVersion: v1
metadata:
  name: compute-resources
  namespace: deleteme
spec:
  hard:
    limits.cpu: null
EOF
resourcequota/compute-resources configured

$ kubectl describe quota
Name:       compute-resources
Namespace:  deleteme
Resource    Used  Hard
--------    ----  ----
```

### What did you expect to happen?

I would expect `kubectl` to be idempotent, and for its behaviour to result in the same end configuration regardless of the initial configuration.

### How can we reproduce it (as minimally and precisely as possible)?

1. Create a new namespace named `deleteme`
2. Run this command:

   ```
   kubectl apply -f - <<EOF
   kind: ResourceQuota
   apiVersion: v1
   metadata:
     name: compute-resources
     namespace: deleteme
   spec:
     hard:
       limits.cpu: null
   EOF
   ```

3. Observe that the ResourceQuota is created and `limits.cpu` is set to `0`
4. Run it again
5. Observe that `limits.cpu` is removed each subsequent time it's run, as long as resourcequota/compute-resource exists
6. Delete resourcequota/compute-resource
7. Run `kubectl apply` again
8. Observe that the ResourceQuota is created and `limits.cpu` is set to `0` again, and so forth
9. Observe the exact same behaviour if `limits.cpu: null` is changed to `limits.cpu:`

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.29.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.26.7+0ef5eae
```

</details>

### Cloud provider

<details>
OpenShift 4.13.12
</details>

### OS version

<details>

```console
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.3 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.3 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
$ uname -a
Linux hostname 6.5.0-17-generic #17~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue Jan 16 14:32:32 UTC 2 x86_64 x86_64 x86_64 GNU/Linux
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by HirazawaUi**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
In #40630, we introduced the ability for patches to remove keys by explicitly setting the null value.

This can lead to inconsistent behavior between create and patch in some scenarios, so to avoid making destructive changes to the Api, we'll do the same on the client side so that kubectl apply on create and kubectl apply on update behave the same way.
#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/123304

#### Does this PR introduce a user-facing change?

```release-note
In the client-side apply on create, defining the null value as "delete the key associated with this value".
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: test/cmd/null.sh [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+run_kubectl_apply_null_tests(){
```
Comment (liggitt) [inline test/cmd/null.sh:21]: nit: let's keep these in test/cmd/apply.sh in run_kubectl_apply_tests

File: test/cmd/null.sh [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    # Create a deployment
+    kubectl create -f hack/testdata/null-propagation/deployment-null.yml "${kube_flags[@]:?}"
```
Comment (liggitt) [inline test/cmd/null.sh:23]: this should be using apply, not create, to create the deployment

File: test/cmd/null.sh [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    # kubectl apply on create should do what kubectl apply on update will accomplish.
+    kubectl apply -f hack/testdata/null-propagation/deployment-null.yml "${kube_flags[@]}"
+    kube::test::get_object_jsonpath_assert "deployment/my-dep-null" "{.spec.template.spec.containers[0].resources.requests.cpu}" ''
+    kube::test::get_object_jsonpath_assert "deployment/my-dep-null" "{.spec.template.spec.containers[0].resources.requests.memory}" '64Mi'
+    kube::test::get_object_jsonpath_assert "deployment/my-dep-null" "{.spec.template.spec.containers[0].terminationMessagePolicy}" 'File'
+
+    kubectl delete deployment my-dep-null "${kube_flags[@]}"
+    # hard.limits.cpu should be nil.
+    kubectl create -f hack/testdata/null-propagation/resourcesquota-null.yml "${kube_flags[@]}"
```
Comment (liggitt) [inline test/cmd/null.sh:38]: this should be using apply, not create, to create the quota

File: test/cmd/null.sh (hunk 2 above)
Comment (HirazawaUi) [reply test/cmd/null.sh]: fixed.

File: test/cmd/null.sh (hunk 3 above)
Comment (HirazawaUi) [reply test/cmd/null.sh]: fixed.

**Comment by HirazawaUi** [comment]
Would you mind **lgtm** it again?
@liggitt

**Pull request merged**